test/test_cpp_api_parity.py

```diff
@@ -0,0 +1,461 @@
+            input_arg_declarations = ',\n'.join(
+                [input_arg_types[i] + ' ' + 'arg' + str(i) for i in range(len(input_arg_types))])
+            input_args = ',\n'.join(['arg' + str(i) for i in range(len(input_arg_types))])
+            test_cpp_sources = template.substitute(
+                module_variant_name=test_params.module_variant_name,
+                module_qualified_name='torch::nn::' + test_params.module_name,
+                cpp_constructor_args=test_params.cpp_constructor_args,
+                input_arg_declarations=input_arg_declarations,
+                input_args=input_args,
+                extra_stmts=extra_stmts)
+            return test_cpp_sources
+
+        def setup_init_test(test_params):
+            # We are generating the attribute equality checks manually here,
+            # because it is not possible to have a `.attributes()` API that returns
+            # non-parameter / non-buffer attributes in a C++ torch::nn module.
+            def generate_attr_equality_checks(module, stmts,
+                                              script_module_prefix='m_init_by_python',
+                                              cpp_module_prefix='m_init_by_cpp'):
+                for name, sub_module in module.named_children():
+                    sub_script_module_prefix = '{}.get_module("{}")'.format(script_module_prefix, name)
+                    sub_cpp_module_prefix = '{}->{}'.format(cpp_module_prefix, name)
+                    generate_attr_equality_checks(sub_module, stmts, sub_script_module_prefix, sub_cpp_module_prefix)
+                for name, param in module._parameters.items():
+                    stmts += CHECK_MODULE_PARAM_EQUALITY.substitute(
+                        script_module_prefix=script_module_prefix,
+                        cpp_module_prefix=cpp_module_prefix,
+                        param_name=name)
+                for name, buffer in module._buffers.items():
+                    stmts += CHECK_MODULE_BUFFER_EQUALITY.substitute(
+                        script_module_prefix=script_module_prefix,
+                        cpp_module_prefix=cpp_module_prefix,
+                        buffer_name=name)
+                for name, attr in module.__dict__.items():
+                    if name not in TORCH_NN_MODULE_IGNORED_ATTRS:
+                        stmts += CHECK_MODULE_ATTR_EQUALITY.substitute(
+                            script_module_prefix=script_module_prefix,
+                            cpp_module_prefix=cpp_module_prefix,
+                            attr_name=name)
+
+            device = test_params.device
+            python_module_class = test_params.python_module_class
+            python_constructor_args = test_params.python_constructor_args
+            example_inputs = test_params.example_inputs
+            has_parity = test_params.has_parity
+
+            with set_has_parity(has_parity):
```

Comment:
Okay, I think I understand why there's a context manager and a global has_parity flag... is it because you can't add `has_parity` as an argument to the module because the cpp module doesn't have a `has_parity` argument?

It took me a while to understand what the context manager is for. One immediate downside to this design is that we can't run these tests in parallel, if we ever chose to.

Would it be better to have a `SampleModuleWithParity` in python / cpp, and then have a `SampleModuleWithoutParity` in python/cpp, and then use those two to do the testing?

Reply:
`has_parity` context manager is removed